keps/sig-node/4800-cpumanager-split-l3cache/README.md

```diff
@@ -0,0 +1,680 @@
+
+#### Story 2
+
+- As a Networking/Telco Engineer, my application is sensitive to uncore cache latency. I want my application to be statically assigned to CPUs that correspond to one single uncore cache so that I can get the maximum throughput from my application. Additionally, I have applications that are also static and of different CPU size requirements. The default Static CPU Manager Policy will result in my applications being assigned to two or more uncore caches. I have a large multi-node Kuberentes cluster. I want a scheduling policy to automatically assign CPUs for the best performance across the fewest amount of uncore cahces based on available CPUs without having to worry about what order/size I deploy my applications so I can ensure I get the best performance possible. 
+
+### Notes/Constraints/Caveats (Optional)
+
+- The name UnCoreCache is directly derived from cAdvisor that is being used as a package. UncoreCache is synonymous L3 cache, or LLC (Last Level Cache). For consistency throughout this document we refer to this as uncore cache. 
+- Using the flow in `CPUManagerPolicy{Alpha,Beta}Options` from [xref](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L110C1-L117C83) which is used to avoid a proliferation of a feature gates. 
+
+### Risks and Mitigations
+
+- Feature is enabled by a static policy option flag.  
+- It does not change behavior of non-static policy
+- It preserves the behavior of other static options.
+
+## Design Details
+
+To make Kubelet be uncore cache aware for CPU sensitive workloads the CPU allocation will require the addition of the `UnCoreCacheID` to enhance CPU scheduling for Pods. This creates the need for updating the `CPUInfo` and the `CPUToplogy` data structures to include `UnCoreCacheID`, which is provided by `cAdvisor` during the CPU discovery process.
+
+A new CPUManager policy option will be introduced, `align-cpus-by-uncorecache`, to enable this feature. With this feature enabled, a modification of the existing default scheduling function is needed to take into consideration of a three level hierarchy: Socket, NUMA, and UnCoreCache. The allocation process will require to attempt to allocate full sockets. Then allocation should attempt the full NUMA nodes. Then allocation must consider the UnCoreCache. Since this policy augments the CPUManager's Static policy, this feature will only be applied to guaranteed pods with whole CPU requests.
```

Comment:
well, again, theoretically, we either taking physical cpu layout priority socket/die/cluster/core/thread model, or here we are following cache hierarchy: L1(thread/single threaded core)/L2(e.g. ARM clusters of cores)/L3 (compute tiles/sockets).

"The allocation process will require to attempt to allocate full sockets. Then allocation should attempt the full NUMA nodes." is a bit unambigous here. NUMA is logical grouping that might differ based on NPS/SNC options in BIOS, thus it can be one or more CPU groups within socket.

Reply:
I can revise it to the socket/die/cluster/core/thread model.

Can you clarify on this:
> NUMA is logical grouping that might differ based on NPS/SNC options in BIOS, thus it can be one or more CPU groups within socket.
I can revise the wording to "NUMA within socket"
Are you referring to the case like NPS0 where the system is dual socket and one numa node on the system?

Reply:
Deleted my previous comment, now I understand.

This may be a dumb question, but in this layout, socket/die/cluster/core/thread, where would NUMA be considered in the hierarchy. The current static scheduler scans through socket and then NUMA.

Reply:
The letter `C` in acronym `NUMA` stands for `CPU`. NUMA first of all, is about memory controller configuration. `static` and in overall TopologyManager design that uses term `NUMA` as the unit of grouping resources is not really correct for modern hardware. `NUMA` split is dependant on BIOS settings for SNC/NPS, and can be equal to `socket` in CPU hierarchy, or it can be between socket/die (or equal to 1 or more dies), depending on design of CPU. In case where details of socket/die/core/thread hierarchy can't be detected properly, we are using other grouping methods: this KEP is talking about using cache layout as a helper, and historically memory controller mode (NUMA) can be used as second helper of detecting hierarchy.

Specifically for paragraph above, we can omit mentioning NUMA at all, or take it into account if single socket contain multiple NUMA groups.